### PR TITLE
De-optionalize Rational and Fraction respelling(numerator:/denominator:)

### LINKED ISSUE #110: Protect Rational with 0 numerator from dividing by 0 in respelling(numerator:)
The `Rational` `respelling(numerator:)` and `respelling(denominator:)` methods use the `/` operator, which in turn uses the `reciprocal` property.

This then turns a `Rational` with zero numerator to one with a zero denominator. Which crashes.

Check if `numerator == 0`, then exit early with an easy operation on the denominator.

### PR DESCRIPTION
**Pull request by mossheim**
Fixes #110. Also does math using doubles rather than floats.

Also, respelling(denominator:) now guards against the case where numerator==0.

### CODE REVIEW

**Comment by mossheim** [comment]
We might want to change the contract/documentation of this function... maybe the asserts should be preconditions? Technically if you're in release mode and the asserts are gone you could get a value that isn't equal to the original Rational. So maybe we write in the contract that you need to be sure you can represent it with that numerator/denominator before proceeding?

File: ArithmeticTools/Rational.swift [hunk 1]
```diff
@@ -100,16 +100,15 @@ public protocol Rational:
     init(_ numerator: Int, _ denominator: Int)
 
     /// - returns: Representation of a `Rational` value with a given `numerator`, if possible.
-    /// Otherwise, `nil`.
     ///
     /// > Preserves the arithmetic value of the original `Rational` value.
-    func respelling(numerator: Int) -> Self?
+    func respelling(numerator: Int) -> Self
 
     /// - returns: Representation of a `Rational` value with a given `denominator`, if
```
Comment (jsbean) [inline ArithmeticTools/Rational.swift:107]: In the Swift stdlib, they use capitalized versions of these labels (`- Returns:`, etc). I've started doing that. Maybe we normalize these as we see them. The lowercased versions still work for the doc gen, but I think we can adopt the standard.

File: ArithmeticTools/Rational.swift [hunk 2]
```diff
@@ -100,16 +100,15 @@ public protocol Rational:
...
-    /// possible. Otherwise, `nil`.
+    /// possible. Assumes the denominator is not 0.
```
Comment (jsbean) [inline ArithmeticTools/Rational.swift:108]: For these, I'd say we can put this expectation within a `- Precondition:` label.

File: ArithmeticTools/Rational.swift [hunk 3]
```diff
@@ -100,16 +100,15 @@ public protocol Rational:
...
     ///
     /// > Preserves the arithmetic value of the original `Rational` value.
-    func respelling(denominator: Int) -> Self?
+    func respelling(denominator: Int) -> Self
```
Comment (jsbean) [inline ArithmeticTools/Rational.swift:111]: Also, is there a better word for this than `respelling` ? This was just an intuitive thing.

Comment (jsbean) [reply ArithmeticTools/Rational.swift:111]: As we are changing the interface here, we might as well question it.

**Pull request merged**